addons/zylann.hterrain/hterrain_detail_layer.gd

```diff
@@ -304,116 +302,84 @@ func _notification(what: int):
-	if terrain == null:
-		_logger.error("Detail layer is not child of a terrain!")
-		return
-	
+func set_terrain_transform(terrain, aabbs: Dictionary):
 	var terrain_transform : Transform = terrain.get_internal_transform()
-
+	var terrain_scale = terrain.map_scale
+	
 	# Update AABBs and transforms, because scale might have changed
-	for k in _chunks:
-		var mmi = _chunks[k]
-		var aabb = _get_chunk_aabb(terrain, Vector3(k.x * CHUNK_SIZE, 0, k.y * CHUNK_SIZE))
+	for k in _mmis:
+		var mmi = _mmis[k]
+		var x_pos = int((k.x * _chunk_size) / terrain_scale.x)
+		var z_pos = int((k.y * _chunk_size) / terrain_scale.z)
+		var aabb = aabbs[Vector2(x_pos, z_pos)]
 		# Nullify XZ translation because that's done by transform already
 		aabb.position.x = 0
 		aabb.position.z = 0
 		mmi.set_aabb(aabb)
-		mmi.set_transform(_get_chunk_transform(terrain_transform, k.x, k.y))
-
-
-func process(delta: float, viewer_pos: Vector3):
-	var terrain = _get_terrain()
-	if terrain == null:
-		_logger.error("DetailLayer processing while terrain is null!")
-		return
+		mmi.set_transform(_get_mmi_transform(terrain_transform, k.x, k.y))
 
+		
+func update(terrain, local_viewer_pos: Vector3, cmin: Vector2, cmax: Vector2, aabbs: Dictionary):
 	if _multimesh_need_regen:
 		_regen_multimesh()
 		_multimesh_need_regen = false
 		# Crash workaround for Godot 3.1
 		# See https://github.com/godotengine/godot/issues/32500
-		for k in _chunks:
-			var mmi = _chunks[k]
+		for k in _mmis:
+			var mmi = _mmis[k]
 			mmi.set_multimesh(_multimesh)
-
-	var local_viewer_pos = terrain.global_transform.affine_inverse() * viewer_pos
-
-	var viewer_cx = local_viewer_pos.x / CHUNK_SIZE
-	var viewer_cz = local_viewer_pos.z / CHUNK_SIZE
-
-	var cr = int(view_distance) / CHUNK_SIZE + 1
-
-	var cmin_x = viewer_cx - cr
-	var cmin_z = viewer_cz - cr
-	var cmax_x = viewer_cx + cr
-	var cmax_z = viewer_cz + cr
-
-	var map_res = terrain.get_data().get_resolution()
-	var map_scale = terrain.map_scale
-
-	var terrain_size_x = map_res * map_scale.x
-	var terrain_size_z = map_res * map_scale.z
-
-	var terrain_chunks_x = terrain_size_x / CHUNK_SIZE
-	var terrain_chunks_z = terrain_size_z / CHUNK_SIZE
-
-	if cmin_x < 0:
-		cmin_x = 0
-	if cmin_z < 0:
-		cmin_z = 0
-	if cmax_x > terrain_chunks_x:
-		cmax_x = terrain_chunks_x
-	if cmax_z > terrain_chunks_z:
-		cmax_z = terrain_chunks_z
-
-	if DEBUG and visible:
+			
+	if DEBUG:
 		_debug_cubes.clear()
-		for cz in range(cmin_z, cmax_z):
-			for cx in range(cmin_x, cmax_x):
-				_add_debug_cube(terrain, _get_chunk_aabb(terrain, Vector3(cx, 0, cz) * CHUNK_SIZE))
-
+	
 	var terrain_transform : Transform = terrain.get_internal_transform()
-
-	for cz in range(cmin_z, cmax_z):
-		for cx in range(cmin_x, cmax_x):
-
+	var terrain_scale = terrain.map_scale
+	
+	for cz in range(cmin.y, cmax.y):
+		for cx in range(cmin.x, cmax.x):
+			var x_pos = int((cx * _chunk_size) / terrain_scale.x)
+			var z_pos = int((cz * _chunk_size) / terrain_scale.z)
```

Comment:
What is x_pos and z_pos here actually?
`cx` and `cz` are chunk positions, which you convert to to cell positions (aka pixel position within heightmap) by multiplying with `_chunk_size`, but then why are you dividing by terrain scale?
I see you use `x_pos` and `y_pos` to index into the `aabbs` dictionary, but chunk coordinates are indexed in grids, coordinates are preferably contiguous. That makes them able to fit in arrays later (instead of dicts). That would maybe even allow to simply use `cx, cz` to find the AABB in the first place, rather than having to do these calculations?
I see we need it in that coordinate space to calculate distance to viewer position, but viewer position can be converted to chunk space before the loop, which eliminates the conversions inside.

Reply:
I updated it to use Vector2(cx, cz) as dictionary key